pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/relocation/SegmentRelocator.java

```diff
@@ -130,40 +137,114 @@ protected void processTable(String tableNameWithType) {
         _externalViewCheckIntervalInMs);
     rebalanceConfig.addProperty(RebalanceConfigConstants.EXTERNAL_VIEW_STABILIZATION_TIMEOUT_IN_MS,
         _externalViewStabilizationTimeoutInMs);
-    // Run rebalance asynchronously
-    _executorService.submit(() -> {
-      try {
-        // Relocating segments to new tiers needs two sequential actions: table rebalance and local tier migration.
-        // Table rebalance moves segments to the new ideal servers, which can change for a segment when its target
-        // tier is updated. New servers can put segments onto the right tier when loading the segments. After that,
-        // all segments are put on the right servers. If any segments are not on their target tier, the server local
-        // tier migration is triggered for them, basically asking the hosting servers to reload them.
-        //
-        // We assume segment target tier is not changed between the two actions, so update target tier here as well,
-        // instead of using a separate task.
-        // TODO: can add some sanity checks on the server side when reloading segments to be more defensive, e.g. only
-        //       migrating segment to new tier when the hosting server is in the server pool configured for that tier.
-        updateTargetTier(tableNameWithType);
+    if (_rebalanceTablesSequentially) {
+      rebalanceTableSequentially(tableNameWithType, tableConfig, rebalanceConfig);
+    } else {
+      _executorService.submit(() -> rebalanceTable(tableNameWithType, tableConfig, rebalanceConfig));
+    }
+  }
 
-        RebalanceResult rebalance =
-            new TableRebalancer(_pinotHelixResourceManager.getHelixZkManager()).rebalance(tableConfig, rebalanceConfig);
-        switch (rebalance.getStatus()) {
-          case NO_OP:
-            LOGGER.info("All segments are already relocated for table: {}", tableNameWithType);
-            migrateToTargetTier(tableNameWithType);
-            break;
-          case DONE:
-            LOGGER.info("Finished relocating segments for table: {}", tableNameWithType);
-            migrateToTargetTier(tableNameWithType);
-            break;
-          default:
-            LOGGER.error("Relocation failed for table: {}", tableNameWithType);
-            break;
+  /**
+   * Use queue to rebalance tables sequentially, meanwhile make sure every table gets its turn.
+   * @param tableNameWithType table requested to rebalance
+   */
+  private void rebalanceTableSequentially(String tableNameWithType, TableConfig tableConfig,
```

Comment:
If I read correctly, if I call this method multiple times, all the tables will be queued, but without one more call, only the first table will be rebalanced.
Instead, I think we should have one background task who keeps polling the tables from the queue and rebalance the table.

Reply:
>  ... one background task who keeps polling ...

You are right. I piggyback on the periodic task scheduler in controller, which schedules SegmentRelocator task at fixed delay, to drive this method.

looks like I can make it more self-driving by checking the waiting queue once a table finishes rebalance
```
_executorService.submit(() -> {
        try {
          rebalanceTable(_tableInProgress, tableConfig, rebalanceConfig);
        } finally {
          LOGGER.debug("Rebalance task completed for table: {}", _tableInProgress);
          _tableInProgress = null; // check if there is table waiting in queue, if so, do rebalanceTable() in current thread. If queue is empty, then leaves the thread.
        }
      });
```

Reply:
actually, will use a blocking queue to simplify this further and avoid the issue pointed out above. stay tuned.